### DISCUSSION TITLE
Perfomance - very slow

### DISCUSSION THREAD

**Discussion by pirex360** [yajra/laravel-datatables#2496] [Performance] [answered]
Hi, first of all, thanks for your great work and for helping buiding better webapps.
Now the problem :
Using serverside, the datatables performance is very slow... i timed my queries, some examples 100ms , but on datatables the render of the table takes 5 to 6 secs.

This example with only 4 records, take 4 seconds to render... also if i sort, or search.
PHP
`````
return DataTables::of(Block::query()->with('doctype')->with('material'))
`````
JS
``````
 $('#datatable').DataTable({
        dom: 'Bfrtip',
        processing: true,
        serverSide: true,
        responsive: true,
        deferRender: true,
...
``````

Where, u can see the screenshot... the time for the queries is <400ms and if i sort any columns ou search... datatable take 4 to 5 seconds to display the result.

![Screenshot 2020-10-26 at 12 21 12](https://user-images.githubusercontent.com/17271768/97171876-0e6ccd00-1786-11eb-90ea-754d62712041.png)

I also post similar issue/problem here : https://github.com/yajra/laravel-datatables/issues/2482#issuecomment-716250549

**Comment by schonhoff** [comment]
Hello,

maybe it is caused through your ```deferRender: true``` option? In the documentation it says:
`this operation can take a not-insignificant amount of time`
https://datatables.net/reference/option/deferRender

I have a server side loading of 5k database rows and my table with 100 entries every page is drawn in under 1 second. And I'm currently at over 1k queries in that time.

**Comment by pirex360** [comment]
Thanks for replying.
In my case, using `deferRender : true` is better... about 0.5s .

My bet is something about the way i use the query `DataTables::of(Block::query()->with('doctype')->with('material'))`

**Comment by schonhoff** [comment]
Maybe try
```
DataTables::of(Block::with(['doctype', 'material']))
```
It is an Eloquent Query but works fine for me.

**Comment by yajra** [comment]
@pirex360 try specifying the the columns that you are using. I usually encounter this slow response when I include a binary field on the selected fields. 

```php
DataTables::of(Block::with(['doctype', 'material'])->select('id', 'doctype_id', 'material_id', 'etc...');
```

**Comment by pirex360** [comment]
Hi, thanks for your effort, but still the performance is bad. The speed improves a bit in production, but still is a bad user experience.

I change the query to. : 
`````
Block::select('id','ndoc','date','blockcode',
                                            '_length','_height','_width','_volume','_weight',
                                            'length','height','width','volume','weight',
                                            'client_id','material_id','doctype_id','supplier_id','color_id','rawmaterial_id')
                                    ->with(['doctype','material','color','rawmaterial']
`````

**Comment by yajra** [comment]
Were you able to inspect the sql execution time? That might give you a clue where the issue is. It could also be on the eager loaded relations. On some codes, I also select the columns when eager loading.

```php
    public function query(Document $model)
    {
        return $model->newQuery()
                     ->select($model->getTable() . '.*')
                     ->when(request('company_id'), function ($q) {
                         $q->where('company_id', request('company_id'));
                     })
                     ->when(request('entity_id'), function ($q) {
                         $q->where('entity_id', request('entity_id'));
                     })
                     ->with('company:id,name')
                     ->with('entity:id,name')
                     ->with('creator:id,first_name,last_name')
                     ->with('updater:id,first_name,last_name');
    }
```

**Comment by pirex360** [comment]
Yes, the sql execution time is only 300ms. 
The 4 seconds  added is from dataTables. 

I will try select columns when eager loading, but they are small tables, only with ID,Name and Timestamps, and max 50 rows.

**Comment by schonhoff** [comment]
Do you have some rendering functions in the frontend?

**Comment by pirex360** [comment]
Hi there, i tried a lot of options... but still the results are slow.

In the frontend

```
$('#datatable').DataTable({
        dom: 'Bfrtip',
        processing: true,
        serverSide: true,
        responsive: true,
        deferRender: false,
        lengthMenu: dtpag,
        language: dtlang_pt,
        stateSave: true,
        ajax: '{!! route('block.dt') !!}',
        autoWidth: false,
        colReorder: true,
        buttons: [
            {
                extend: 'colvis',
                className: 'btn-info btn-sm rounded-0',
                collectionLayout: 'three-column',
                text: "<i class='fas fa-eye'></i> Ver/Ocultar Colunas",
                columns: ':not(.noVis)'
            },
            {
                className: 'ml-1 btn-info btn-sm rounded-0',
                text: "<i class='fas fa-grip-lines-vertical'></i> Repor Ordem",
                action: function ( e, dt, node, config ) {
                    dt.colReorder.reset();
                }
            },
            {
                className: 'ml-1 btn-info btn-sm rounded-0',
                text: "<i class='uil-refresh'></i> Repor Tabela",
                action: function ( e, dt, node, config ) {
                    dt.state.clear();
                    window.location.reload();
                }
            }
        ],
        columns: [
            { data: 'date', name: 'date', orderable: true, searchable: false},
            { data: 'blockcode', name: 'blockcode', orderable: true, searchable: true, visible: true },
            { data: 'material.name', name: 'material.name', className: 'text-left', orderable: true, searchable: true},
            { data: 'color.name', name: 'color.name', className: 'text-left', orderable: true, searchable: true},
            { data: 'rawmaterial.name', name: 'rawmaterial.name', className: 'text-left', orderable: true, searchable: true},
            { data: '_supplier', name: '_supplier', className: 'text-left', orderable: false, searchable: true, width: '100%',visible: false},
            { data: '_client', name: '_client', className: 'text-left', orderable: false, searchable: true, width: '100%', visible: false},
            { data: 'ndoc', name: 'ndoc', className: 'text-left', orderable: true, searchable: true, width: '100%', visible:false},
            { data: 'doctype.name', name: 'doctype.name', className: 'text-left', orderable: true, searchable: true, width: '100%', visible: false},
            { data: '_length', name: '_length', className: 'text-center', orderable: true, searchable: true},
            { data: '_height', name: '_height', className: 'text-center', orderable: true, searchable: true},
            { data: '_width', name: '_width', className: 'text-center', orderable: true, searchable: true},
            { data: '_volume', name: '_volume', className: 'text-center', orderable: true, searchable: true},
            { data: '_weight', name: '_weight', className: 'text-center', orderable: true, searchable: true},
            { data: 'length', name: '_length', className: 'text-center', orderable: true, searchable: true},
            { data: 'height', name: '_height', className: 'text-center', orderable: true, searchable: true},
            { data: 'width', name: '_width', className: 'text-center', orderable: true, searchable: true},
            { data: 'volume', name: '_volume', className: 'text-center', orderable: true, searchable: true},
            { data: 'weight', name: 'weight', className: 'text-center', orderable: true, searchable: true},

            { data: 'options', name: 'options', className: 'noVis', orderable: false, searchable: false },
        ],
        order: [[1, 'asc']],
        drawCallback: function(settings) {
            // SweetAlert
            $('.btn-del').click(function (event) {
                    let id = event.target.id.substring(4);
                    let recordName = $('#'+id).data("name");
                    let msg = `Bloco nº. <b>${recordName}</b> será removido permanentemente`;

                    Swal.fire({
                        title: "Deseja Apagar ?",
                        html: msg,
                        icon: "warning",
                        showCancelButton: true,
                        cancelButtonText: "Cancelar",
                        confirmButtonColor: "#34c38f",
                        cancelButtonColor: "#f46a6a",
                        confirmButtonText: "Sim, Apagar!"
                        }).then(function (result) {
                            if (result.value) {
                                // action ok
                                let url = "{{ route('block.delete', '') }}"+"/"+id;
                                location.href=url;
                            }
                        });
            });
            // SweetAlert

        }

    });

```

And the Service Logic

 ```
return DataTables::of( Block::with(['doctype','material','color','rawmaterial'])
                                        ->select('blocks.id','ndoc','date','blockcode',
                                            '_length','_height','_width','_volume','_weight',
                                            'length','height','width','volume','weight',
                                            'client_id','material_id','doctype_id','supplier_id','color_id','rawmaterial_id')
        )

        ->setRowId(function ($data) {
            return $data->id;
        })

        ->setRowAttr([
            'data-name' => '{{$blockcode}}',
        ])

        ->addColumn('_supplier', function($data){
            $relation = $this->suppliers->firstWhere('no', $data->supplier_id);
            $field = $relation['nome'];

            return $field;
        })

        ->addColumn('_client', function($data){
            $relation = $this->clients->firstWhere('no', $data->client_id);
            $field = $relation['nome'];

            return $field;
        })

        ->addColumn('options', function ($data) {
            $editUrl = route('block.edit', ['id' => $data->id]);
            $viewUrl = route('block.show', ['id' => $data->id]);

            $btn_view = "
            <a href='{$viewUrl}' class='dropdown-item text-left font-weight-bold font-size-13' id='view-{$data->id}'>
                <i class='fas fa-eye text-primary mr-2'></i>". __('Ver') . "
            </a>";
            $btn_edit = "<a href='{$editUrl}' class='dropdown-item text-left font-weight-bold font-size-13' id='edit-{$data->id}'>
                <i class='fas fa-edit text-info mr-2'></i>". __('Editar') .
            "</a>";
            $btn_delete = "
            <a href='javascript:void(0);' class='dropdown-item text-left font-weight-bold font-size-13 btn-del' id='del-{$data->id}'>
                <i class='fas fa-trash text-danger mr-2'></i>". __('Apagar') .
            "</a>";
            // Button Auth Logic
            $btn_view = ((Auth::user()->can('block-view')) ? $btn_view : NULL);
            $btn_edit = ((Auth::user()->can('block-edit')) ? $btn_edit : NULL);
            $btn_delete = ((Auth::user()->can('block-delete')) ? $btn_delete : NULL);

            $button_dropdown = "
            <div class='btn-group dropleft '>
                <button class='btn btn-sm btn-info rounded-0 waves-effect dropdown-toggle' type='button' id='dropdownMenuButton' data-toggle='dropdown' aria-haspopup='true' aria-expanded='false'>
                    <i class='mdi mdi-chevron-left'></i> <i class='uil-cog'></i>
                </button>
                <div class='dropdown-menu drop-menu-left border-info rounded-0 mr-2 mt-0 mb-0 pb-0' style='border:1px solid;box-shadow:2px 2px 5px grey;' >
                    <div class='bg-info text-right font-size-12 font-weight-bold text-light p-1' style='margin-top:-8px;'>
                    <span class='mr-2'>Nº. Bloco: {$data->blockcode}</span>
                    </div>
                    $btn_view
                    $btn_edit
                    $btn_delete
                </div>
            </div>
            ";

            $buttons = $button_dropdown;
            return $buttons;
        })

        ->editColumn('date', function($data){
            $formatedDate = Carbon::createFromFormat('Y-m-d', $data->date)->format('d-m-Y');
            return $formatedDate;
        })


        ->rawColumns(['options'])

        ->make(true);
```

**Comment by yajra** [comment]
Not sure but it seems like a N+1 issue? Try removing these 2 added columns.

```php
       ->addColumn('_supplier', function($data){
           $relation = $this->suppliers->firstWhere('no', $data->supplier_id);
           $field = $relation['nome'];

           return $field;
       })

       ->addColumn('_client', function($data){
           $relation = $this->clients->firstWhere('no', $data->client_id);
           $field = $relation['nome'];

           return $field;
       })
```

**Comment by pirex360** [comment]
Hi, i just gained about 120ms in 5s, this colums are collections from other database that are fetched via api data of the clients and suppliers, the collections are global variables of the class. So i just need to fetch in the constructor and then i just compare the id to fetch the name.

I really worried about this scenario, because all my project is based on dataTables for UI of data. The problem, is that the time of table sorting, table render, search is very similar with 20 records and 2500 records, i dont think it's a query issue... but some DataTables configuration or optimization that i missed.

**Comment by yajra** [comment]
Hmm, usual cause of slow performance is due to collection, N+1 and slow query count. Can you try removing all the add & edit columns and observe the performance? The issue might not be on the package but by other factors.

Also, do you have laravel-debugbar? You can debug further using that package.

**Comment by pirex360** [comment]
For local i use devilbox in a Mac OSX... , i will try in the production server.
Yes i use the laravel-debugbar, this is the reason i see the sql queries times.

Thanks for all the support, i will keep up seaching and update this.

**Comment by schonhoff** [comment]
Can you remove the drawCallback function just for a test? I had the same issue as I started with DataTables and on my end it was a callback function from DataTables that caused the issue on long client side rendering.

**Comment by pirex360** [comment]
Ok, i will try without drawcallback! I will report soon.

I was in a fight with the eloquent query, thinking that has a problem N+1, but the debugbar show me the queries and the times. 
Has you can see, they are fast ... the processing of the table takes for this example near 4seconds... i dont understand why.

`"queries": [
    {
      "query": "select count(*) as aggregate from (select '1' as `row_count` from `blocks`) count_row_table",
      "bindings": [
        
      ],
      "time": "29.7"
    },
    {
      "query": "select `blocks`.`id`, `ndoc`, `date`, `blockcode`, `_length`, `_height`, `_width`, `_volume`, `_weight`, `length`, `height`, `width`, `volume`, `weight`, `client_id`, `material_id`, `doctype_id`, `supplier_id`, `color_id`, `rawmaterial_id`, `blockstate_id` from `blocks` order by `blockcode` desc limit 10 offset 0",
      "bindings": [
        
      ],
      "time": "1.56"
    },
    {
      "query": "select `id`, `name` from `doctypes` where `doctypes`.`id` in (2)",
      "bindings": [
        
      ],
      "time": "35.25"
    },
    {
      "query": "select `id`, `name` from `materials` where `materials`.`id` in (2, 4)",
      "bindings": [
        
      ],
      "time": "2.71"
    },
    {
      "query": "select `id`, `name` from `colors` where `colors`.`id` in (5, 8)",
      "bindings": [
        
      ],
      "time": "35.72"
    },
    {
      "query": "select `id`, `name` from `rawmaterials` where `rawmaterials`.`id` in (47, 76)",
      "bindings": [
        
      ],
      "time": "1.79"
    },
    {
      "query": "select `id`, `name` from `blockstates` where `blockstates`.`id` in (1, 2)",
      "bindings": [
        
      ],
      "time": "35.95"
    }
  ],`

**Comment by pirex360** [comment, marked as answer]
Finally the end of this mini-nightmare. I hope the thread will help others.

When i install the debugbar, i notice a large number of "GATE" request, because in the ->addColumn i use 3 checks permissions (spatie) , if user `@can('xxxx')` for 3 buttons (view, edit, delete) - so every line , 3x Gate Requests where made.

**Solution :** before datatables init -> check only one time the permissions of the user... and use the same logic for every line, but using a local variable with the roles of the current user, because before : 3x Gate Calls ``@can('xxxx')`` where made per line ( i think equals 3 queries extra per line ), in 2500 records... imagine the lag.

**Thanks for all the help and the support.**

**Reply by yajra** [reply]
Converted the issue to discussion and marking this as the answer. This is a N+1 issue due to multiple gate calls (But I think it should be cached by spatie to prevent N+1). Thanks!
